### PR TITLE
Add New Security Script Settings

### LINKED ISSUE #24532: Simplify scripting enabled settings
We currently have 3 concepts in scripts: language, type (inline, file, stored), and context. We also have "fine grained control" over the various combinations of these that can be allowed. These settings are automatically added in the script service, and the logic is complex with fallbacks.

As we are working towards a simpler scripting api, with contexts actually implying the variables available to a script, we should simplify these settings. I propose to have each of these concepts allow for disabling, but not the endless combinations; that level of control is simply not necessary.

My proposed settings are:
```
script.types_allowed: ['inline']
script.contexts_allowed: ['search_template', 'whatever']
```

The plan would be to add these alternative settings in 5.5, at the same time as disabling the old fine grained settings.

### PR DESCRIPTION
**Pull request by jdconrad**
Adds allowed_types and allowed_contexts as new security settings for scripts described in detail in the issue #24532.

Closes #24532

### CODE REVIEW

**Comment by jdconrad** [comment]
@elasticmachine please test this

**Comment by jdconrad** [comment]
@rjernst If you're feeling up to it, I would appreciate a review on this :)

File: core/src/main/java/org/elasticsearch/script/ScriptModes.java [hunk 1]
```diff
@@ -38,12 +44,55 @@
...
+    private final Set<String> contextsAllowed;
+
+    ScriptModes(ScriptContextRegistry scriptContextRegistry, ScriptSettings scriptSettings, Settings settings) {
         HashMap<String, Boolean> scriptModes = new HashMap<>();
         for (Setting<Boolean> scriptModeSetting : scriptSettings.getScriptLanguageSettings()) {
             scriptModes.put(scriptModeSetting.getKey(), scriptModeSetting.get(settings));
         }
         this.scriptEnabled = Collections.unmodifiableMap(scriptModes);
+
+        typesAllowed = TYPES_ALLOWED_SETTING.exists(settings) ? new HashSet<>() : null;
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/script/ScriptModes.java]: I think in the future we can simplify this a bit, so the Setting value is not a List, but that can be in future work as I know this class will probably go away in master with context work.

File: docs/reference/modules/scripting/security.asciidoc [hunk 2]
```diff
@@ -87,6 +87,51 @@ change from the defaults described above. You should be very, very careful
 when allowing more than the defaults. Any extra permissions weakens the total
 security of the Elasticsearch deployment.
 
+[[allowed-script-types-setting]]
+[float]
+=== Allowed script types setting
+
+By default all script types are allowed to be executed.  This can modified using the
```
Comment (rjernst) [inline docs/reference/modules/scripting/security.asciidoc]: can -> can be

File: docs/reference/modules/scripting/security.asciidoc [hunk 3]
```diff
@@ -87,6 +87,51 @@ change from the defaults described above. You should be very, very careful
...
+setting `script.allowed_types`.  Should this setting be used, only the types specified
```
Comment (rjernst) [inline docs/reference/modules/scripting/security.asciidoc]: I think you can remove the "Should this setting be used" part.

**Review by rjernst — APPROVED** [review]
LGTM, I left a few minor comments.

File: core/src/main/java/org/elasticsearch/script/ScriptModes.java (hunk 1 above)
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/script/ScriptModes.java]: Agree completely on this.

File: core/src/main/java/org/elasticsearch/script/ScriptContextRegistry.java [hunk 4]
```diff
@@ -64,6 +64,13 @@ public ScriptContextRegistry(Collection<ScriptContext.Plugin> customScriptContex
...
      * @return <tt>true</tt> if the provided {@link ScriptContext} is supported, <tt>false</tt> otherwise
      */
+    boolean isSupportedContext(String scriptContext) {
+        return scriptContexts.containsKey(scriptContext);
+    }
+
+    /**
+     * @return <tt>true</tt> if the provided {@link ScriptContext} is supported, <tt>false</tt> otherwise
+     */
     boolean isSupportedContext(ScriptContext scriptContext) {
```
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/script/ScriptContextRegistry.java]: Done.  There were only two callers and one was a test.

File: docs/reference/modules/scripting/security.asciidoc (hunk 2 above)
Comment (jdconrad) [inline docs/reference/modules/scripting/security.asciidoc]: Fixed.

File: docs/reference/modules/scripting/security.asciidoc (hunk 3 above)
Comment (jdconrad) [inline docs/reference/modules/scripting/security.asciidoc]: Done.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review!  As soon as the CI build passes I will commit.

**Pull request merged**